File: scripts/changelog-generation/validate-provenance.js
'use strict';
/**
 * validate-provenance.js
 *
 * Validates the agent's changelog markdown and provenance JSON against the
 * evidence manifest to ensure every changelog bullet is backed by a real PR
 * present in the evidence.
 *
 * Usage (CLI):
 *   node scripts/changelog-generation/validate-provenance.js \
 *     --evidence  memory/changelog-generation/evidence.json \
 *     --provenance memory/changelog-generation/provenance.json \
 *     --changelog  CHANGELOG.md \
 *     [--section-header "## [Unreleased]"|"## [0.14.4] - 2026-04-16"]
 *
 * Env vars (alternative to CLI args):
 *   EVIDENCE_PATH, PROVENANCE_PATH, CHANGELOG_PATH, SECTION_HEADER
 *
 * Exits 0 on success, 1 on validation failure.
 * Outputs a JSON result: { valid: boolean, errors: string[], warnings: string[] }
 *
 * Exports: validateProvenance (for unit testing)
 */

const fs = require('node:fs');
const path = require('node:path');

// ---------------------------------------------------------------------------
// Heuristics
// ---------------------------------------------------------------------------

/** Regex that matches a 40-character hex SHA (commit hash). */
const COMMIT_SHA_RE = /\b[0-9a-f]{40}\b/i;

/** Regex that matches a short (7-12 char) hex SHA. */
const SHORT_SHA_RE = /\b[0-9a-f]{7,12}\b/i;

/**
 * Extracts all #NNN PR number references from a block of markdown text.
 * Only matches references of the form #NNN (not inside URLs).
 *
 * @param {string} text
 * @returns {number[]}
 */
function extractPRReferences(text) {
  // Match #NNN not preceded by / (which would be a URL path)
  const matches = [...text.matchAll(/(?<!\/)#(\d+)/g)];
  return [...new Set(matches.map((m) => parseInt(m[1], 10)))];
}

/**
 * Extracts bullet lines from a changelog section block (the part between
 * the section header and the next ## header).
 *
 * @param {string} sectionContent — text of the section (may include header)
 * @returns {string[]}
 */
function extractBulletLines(sectionContent) {
  return sectionContent
    .split('\n')
    .filter((line) => /^\s*[*-]\s/.test(line));
}

/**
 * Detects whether a line looks like commit-level narration:
 *   - Contains a full 40-char SHA
 *   - Is a short generic sentence without any #NNN PR reference and
 *     contains a 7-12 char hex string
 *
 * @param {string} line
 * @returns {boolean}
 */
function looksLikeCommitNarration(line) {
  if (COMMIT_SHA_RE.test(line)) return true;
  // Short hex SHA without PR reference is suspicious
  if (SHORT_SHA_RE.test(line) && !/#\d+/.test(line)) return true;
  return false;
}

// ---------------------------------------------------------------------------
// Core validation
// ---------------------------------------------------------------------------

/**
 * Validates provenance JSON and changelog markdown against the evidence manifest.
 *
 * @param {object} params
 * @param {object} params.evidence   — parsed evidence manifest JSON
 * @param {object} params.provenance — parsed provenance JSON
 * @param {string} params.changelogSection — the raw markdown of the target section
 *   (including the section header line, e.g. "## [Unreleased]\n...")
 * @param {string} [params.expectedHeader] — e.g. "## [Unreleased]" or "## [0.14.4] - 2026-04-16"
 *   If omitted, derived from evidence.target_section.
 * @returns {{ valid: boolean, errors: string[], warnings: string[] }}
 */
function validateProvenance({ evidence, provenance, changelogSection, expectedHeader }) {
  const errors = [];
  const warnings = [];

  // Build a set of known PR numbers from the evidence manifest
  const evidencePRNumbers = new Set(
    (evidence.pull_requests ?? []).map((pr) => pr.number)
  );

  // --- Check 1: provenance bullets map to known PRs ---
  const bullets = provenance.bullets ?? [];
  for (const bullet of bullets) {
    const prNumbers = bullet.pr_numbers ?? [];
    if (prNumbers.length === 0) {
      errors.push(
        `Provenance bullet has no pr_numbers: "${bullet.text ?? '(no text)'}"`
      );
    }
    for (const prNum of prNumbers) {
      if (!evidencePRNumbers.has(prNum)) {
        errors.push(
          `Provenance bullet references PR #${prNum} which is NOT in the evidence manifest: "${bullet.text ?? ''}"`
        );
      }
    }
  }

  // --- Check 2: every #NNN in the changelog markdown is backed by evidence ---
  const changelogPRRefs = extractPRReferences(changelogSection);
  for (const prNum of changelogPRRefs) {
    if (!evidencePRNumbers.has(prNum)) {
      errors.push(
        `Changelog references PR #${prNum} which is NOT in the evidence manifest (fabricated reference?)`
      );
    }
  }

  // --- Check 3: no commit-level narration ---
  const bulletLines = extractBulletLines(changelogSection);
  for (const line of bulletLines) {
    if (looksLikeCommitNarration(line)) {
      errors.push(
        `Changelog bullet appears to contain commit-level narration (SHA found): "${line.trim()}"`
      );
    }
    // Warn about bullets without any PR reference
    if (!/#\d+/.test(line)) {
      warnings.push(
        `Changelog bullet has no PR reference (#NNN): "${line.trim()}"`
      );
    }
  }

  // --- Check 4: section header matches expected ---
  const resolvedExpectedHeader = expectedHeader ?? evidence.target_section ?? '';
  if (resolvedExpectedHeader) {
    const lines = changelogSection.split('\n');
    const firstNonEmpty = lines.find((l) => l.trim() !== '');
    if (firstNonEmpty && !firstNonEmpty.startsWith(resolvedExpectedHeader)) {
      errors.push(
        `Changelog section header "${firstNonEmpty.trim()}" does not match expected "${resolvedExpectedHeader}"`
      );
    }
  }

  // --- Check 5: markdown format — bullets use "- " or "* " format ---
  for (const line of bulletLines) {
    if (!/^\s*[-*]\s/.test(line)) {
      warnings.push(`Unexpected bullet format (expected "- " or "* "): "${line.trim()}"`);
    }
  }

  const valid = errors.length === 0;
  return { valid, errors, warnings };
}

/**
 * Reads and parses a JSON file. Returns the parsed object.
 *
 * @param {string} filePath
 * @returns {object}
 */
function readJSON(filePath) {
  const raw = fs.readFileSync(filePath, 'utf8');
  return JSON.parse(raw);
}

/**
 * Extracts a named section from CHANGELOG.md content.
 * Returns the content from the section header up to (but not including) the next ## header.
 *
 * @param {string} changelogContent
 * @param {string} header — e.g. "## [Unreleased]" or "## [0.14.4]"
 * @returns {string|null}
 */
function extractSectionFromChangelog(changelogContent, header) {
  const lines = changelogContent.split('\n');
  let inSection = false;
  const sectionLines = [];

  for (const line of lines) {
    if (!inSection) {
      if (line.startsWith(header)) {
        inSection = true;
        sectionLines.push(line);
      }
    } else {
      // Stop at the next ## header (but not the same one)
      if (line.startsWith('## ') && !line.startsWith(header)) {
        break;
      }
      sectionLines.push(line);
    }
  }

  return inSection ? sectionLines.join('\n') : null;
}

// ---------------------------------------------------------------------------
// CLI entry point
// ---------------------------------------------------------------------------

if (require.main === module) {
  // Parse CLI args
  const args = process.argv.slice(2);
  const argMap = {};
  for (let i = 0; i < args.length; i += 2) {
    const key = args[i].replace(/^--/, '').replace(/-([a-z])/g, (_, c) => c.toUpperCase());
    argMap[key] = args[i + 1];
  }

  const evidencePath =
    argMap.evidence ||
    process.env.EVIDENCE_PATH ||
    'memory/changelog-generation/evidence.json';

  const provenancePath =
    argMap.provenance ||
    process.env.PROVENANCE_PATH ||
    'memory/changelog-generation/provenance.json';

  const changelogPath =
    argMap.changelog || process.env.CHANGELOG_PATH || 'CHANGELOG.md';

  const sectionHeaderArg =
    argMap.sectionHeader || process.env.SECTION_HEADER || '';

  let evidence, provenance, changelogContent;

  try {
    evidence = readJSON(evidencePath);
  } catch (err) {
    const result = { valid: false, errors: [`Cannot read evidence manifest: ${err.message}`], warnings: [] };
    process.stdout.write(JSON.stringify(result, null, 2) + '\n');
    process.exit(1);
  }

  try {
    provenance = readJSON(provenancePath);
  } catch (err) {
    const result = { valid: false, errors: [`Cannot read provenance file: ${err.message}`], warnings: [] };
    process.stdout.write(JSON.stringify(result, null, 2) + '\n');
    process.exit(1);
  }

  try {
    changelogContent = fs.readFileSync(changelogPath, 'utf8');
  } catch (err) {
    const result = { valid: false, errors: [`Cannot read CHANGELOG.md: ${err.message}`], warnings: [] };
    process.stdout.write(JSON.stringify(result, null, 2) + '\n');
    process.exit(1);
  }

  const targetHeader = sectionHeaderArg || evidence.target_section || '## [Unreleased]';
  // Extract just the section heading prefix for matching (strip date part for lookup)
  const headerPrefix = targetHeader.split(' - ')[0];
  const changelogSection = extractSectionFromChangelog(changelogContent, headerPrefix);

  if (!changelogSection) {
    const result = {
      valid: false,
      errors: [`Section "${targetHeader}" not found in ${changelogPath}`],
      warnings: [],
    };
    process.stdout.write(JSON.stringify(result, null, 2) + '\n');
    process.exit(1);
  }

  const result = validateProvenance({
    evidence,
    provenance,
    changelogSection,
    expectedHeader: targetHeader,
  });

  process.stdout.write(JSON.stringify(result, null, 2) + '\n');

  if (!result.valid) {
    process.stderr.write(`Validation failed: ${result.errors.length} error(s)\n`);
    process.exit(1);
  }

  process.stderr.write(`Validation passed${result.warnings.length > 0 ? ` (${result.warnings.length} warning(s))` : ''}\n`);
  process.exit(0);
}

module.exports = {
  classifyPR: undefined, // re-exported for convenience from build-evidence-manifest
  validateProvenance,
  extractPRReferences,
  extractBulletLines,
  looksLikeCommitNarration,
  extractSectionFromChangelog,
  readJSON,
};

File: scripts/changelog-generation/validate_provenance.go
// Licensed to Elasticsearch B.V. under one or more contributor
// license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright
// ownership. Elasticsearch B.V. licenses this file to you under
// the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package main

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

// ---------------------------------------------------------------------------
// Regex patterns
// ---------------------------------------------------------------------------

// commitSHARE matches a 40-character hex SHA (commit hash).
var commitSHARE = regexp.MustCompile(`(?i)\b[0-9a-f]{40}\b`)

// shortSHARE matches a short (7-12 char) hex SHA.
var shortSHARE = regexp.MustCompile(`(?i)\b[0-9a-f]{7,12}\b`)

// prRefRE matches #NNN not preceded by / (which would be a URL path segment).
// We simulate the negative lookbehind by extracting the full match and checking the preceding byte.
var prRefRE = regexp.MustCompile(`(^|[^/])#(\d+)`)

// ---------------------------------------------------------------------------
// Core helper functions
// ---------------------------------------------------------------------------

// extractPRReferences extracts all #NNN PR number references from markdown text,
// excluding references that appear after a / (URL paths).
func extractPRReferences(text string) []int {
	matches := prRefRE.FindAllStringSubmatch(text, -1)
	seen := map[int]bool{}
	var result []int
	for _, m := range matches {
		n, err := strconv.Atoi(m[2])
		if err != nil {
			continue
		}
		if !seen[n] {
			seen[n] = true
			result = append(result, n)
		}
	}
	return result
}

// extractBulletLines extracts bullet lines from a changelog section.
var bulletLineRE = regexp.MustCompile(`^\s*[*-]\s`)

func extractBulletLines(sectionContent string) []string {
	var bullets []string
	for line := range strings.SplitSeq(sectionContent, "\n") {
		if bulletLineRE.MatchString(line) {
			bullets = append(bullets, line)
		}
	}
	return bullets
}

// looksLikeCommitNarration returns true if a line appears to reference a commit SHA.
var prNumInlineRE = regexp.MustCompile(`#\d+`)

func looksLikeCommitNarration(line string) bool {
	if commitSHARE.MatchString(line) {
		return true
	}
	// Short hex SHA without PR reference is suspicious
	if shortSHARE.MatchString(line) && !prNumInlineRE.MatchString(line) {
		return true
	}
	return false
}

// extractSectionFromChangelog extracts a named section from CHANGELOG.md content.
// Returns the content from the section header up to (but not including) the next ## header.
// Uses exact prefix matching: header must equal the line, or line starts with header+" " or header+"\t".
func extractSectionFromChangelog(changelogContent, header string) (string, bool) {
	lines := strings.Split(changelogContent, "\n")
	inSection := false
	var sectionLines []string

	headerMatches := func(line string) bool {
		return line == header ||
			strings.HasPrefix(line, header+" ") ||
			strings.HasPrefix(line, header+"\t")
	}

	for _, line := range lines {
		if !inSection {
			if headerMatches(line) {
				inSection = true
				sectionLines = append(sectionLines, line)
			}
		} else {
			// Stop at the next ## header (but not the same one)
			if strings.HasPrefix(line, "## ") && !headerMatches(line) {
				break
			}
			sectionLines = append(sectionLines, line)
		}
	}

	if !inSection {
		return "", false
	}
	return strings.Join(sectionLines, "\n"), true
}

// ---------------------------------------------------------------------------
// JSON input types
// ---------------------------------------------------------------------------

type evidenceManifest struct {
	TargetSection string       `json:"target_section"`
	PullRequests  []evidencePR `json:"pull_requests"`
}

type evidencePR struct {
	Number int    `json:"number"`
	Title  string `json:"title"`
	URL    string `json:"url"`
}

type provenanceFile struct {
	Bullets []provenanceBullet `json:"bullets"`
}

type provenanceBullet struct {
	Text      string `json:"text"`
	PRNumbers []int  `json:"pr_numbers"`
}

// ---------------------------------------------------------------------------
// Core validation
// ---------------------------------------------------------------------------

type validationResult struct {
	Valid    bool     `json:"valid"`
	Errors   []string `json:"errors"`
	Warnings []string `json:"warnings"`
}

type validateProvenanceParams struct {
	evidence         *evidenceManifest
	provenance       *provenanceFile
	changelogSection string
	expectedHeader   string
}

func validateProvenance(p validateProvenanceParams) validationResult {
	var errors []string
	var warnings []string

	// Build a set of known PR numbers from the evidence manifest
	evidencePRNumbers := map[int]bool{}
	for _, pr := range p.evidence.PullRequests {
		evidencePRNumbers[pr.Number] = true
	}

	// --- Check 1: provenance bullets map to known PRs ---
	bullets := p.provenance.Bullets
	for _, bullet := range bullets {
		if len(bullet.PRNumbers) == 0 {
			text := bullet.Text
			if text == "" {
				text = "(no text)"
			}
			errors = append(errors, fmt.Sprintf("Provenance bullet has no pr_numbers: %q", text))
		}
		for _, prNum := range bullet.PRNumbers {
			if !evidencePRNumbers[prNum] {
				errors = append(errors, fmt.Sprintf(
					"Provenance bullet references PR #%d which is NOT in the evidence manifest: %q",
					prNum, bullet.Text,
				))
			}
		}
	}

	// --- Check 2: every #NNN in the changelog markdown is backed by evidence ---
	changelogPRRefs := extractPRReferences(p.changelogSection)
	for _, prNum := range changelogPRRefs {
		if !evidencePRNumbers[prNum] {
			errors = append(errors, fmt.Sprintf(
				"Changelog references PR #%d which is NOT in the evidence manifest (fabricated reference?)",
				prNum,
			))
		}
	}

	// --- Check 3: no commit-level narration ---
	bulletLines := extractBulletLines(p.changelogSection)
	for _, line := range bulletLines {
		if looksLikeCommitNarration(line) {
			errors = append(errors, fmt.Sprintf(
				"Changelog bullet appears to contain commit-level narration (SHA found): %q",
				strings.TrimSpace(line),
			))
		}
		// Reject bullets without any PR reference
		if !prNumInlineRE.MatchString(line) {
			errors = append(errors, fmt.Sprintf(
				"Changelog bullet has no PR reference (#NNN): %q",
				strings.TrimSpace(line),
			))
		}
	}

	// --- Check 4: section header matches expected ---
	resolvedExpectedHeader := p.expectedHeader
	if resolvedExpectedHeader == "" {
		resolvedExpectedHeader = p.evidence.TargetSection
	}
	if resolvedExpectedHeader != "" {
		lines := strings.Split(p.changelogSection, "\n")
		var firstNonEmpty string
		for _, l := range lines {
			if strings.TrimSpace(l) != "" {
				firstNonEmpty = l
				break
			}
		}
		if firstNonEmpty != "" && strings.TrimSpace(firstNonEmpty) != resolvedExpectedHeader {
			errors = append(errors, fmt.Sprintf(
				"Changelog section header %q does not match expected %q",
				strings.TrimSpace(firstNonEmpty),
				resolvedExpectedHeader,
			))
		}
	}

	valid := len(errors) == 0
	// Return non-nil slices for clean JSON output
	if errors == nil {
		errors = []string{}
	}
	if warnings == nil {
		warnings = []string{}
	}
	return validationResult{Valid: valid, Errors: errors, Warnings: warnings}
}

// ---------------------------------------------------------------------------
// CLI entry point
// ---------------------------------------------------------------------------

func runValidateProvenance(args []string) {
	// Parse --key value pairs
	argMap := parseArgMap(args)

	evidencePath := argMapGet(argMap, "evidence", os.Getenv("EVIDENCE_PATH"), "/tmp/gh-aw/agent/evidence.json")
	provenancePath := argMapGet(argMap, "provenance", os.Getenv("PROVENANCE_PATH"), "/tmp/gh-aw/agent/provenance.json")
	changelogPath := argMapGet(argMap, "changelog", os.Getenv("CHANGELOG_PATH"), "CHANGELOG.md")
	sectionHeaderArg := argMapGet(argMap, "section-header", os.Getenv("SECTION_HEADER"), "")

	writeResult := func(r validationResult) {
		out, _ := json.MarshalIndent(r, "", "  ")
		fmt.Println(string(out))
	}

	evidenceRaw, err := os.ReadFile(evidencePath)
	if err != nil {
		r := validationResult{Valid: false, Errors: []string{fmt.Sprintf("Cannot read evidence manifest: %s", err.Error())}, Warnings: []string{}}
		writeResult(r)
		os.Exit(1)
	}
	var evidence evidenceManifest
	if err := json.Unmarshal(evidenceRaw, &evidence); err != nil {
		r := validationResult{Valid: false, Errors: []string{fmt.Sprintf("Cannot parse evidence manifest: %s", err.Error())}, Warnings: []string{}}
		writeResult(r)
		os.Exit(1)
	}

	provenanceRaw, err := os.ReadFile(provenancePath)
	if err != nil {
		r := validationResult{Valid: false, Errors: []string{fmt.Sprintf("Cannot read provenance file: %s", err.Error())}, Warnings: []string{}}
		writeResult(r)
		os.Exit(1)
	}
	var provenance provenanceFile
	if err := json.Unmarshal(provenanceRaw, &provenance); err != nil {
		r := validationResult{Valid: false, Errors: []string{fmt.Sprintf("Cannot parse provenance file: %s", err.Error())}, Warnings: []string{}}
		writeResult(r)
		os.Exit(1)
	}

	changelogContent, err := os.ReadFile(changelogPath)
	if err != nil {
		r := validationResult{Valid: false, Errors: []string{fmt.Sprintf("Cannot read CHANGELOG.md: %s", err.Error())}, Warnings: []string{}}
		writeResult(r)
		os.Exit(1)
	}

	targetHeader := sectionHeaderArg
	if targetHeader == "" {
		targetHeader = evidence.TargetSection
	}
	if targetHeader == "" {
		targetHeader = headerUnreleased
	}

	// Extract just the section heading prefix for matching (strip date part for lookup)
	headerPrefix := strings.SplitN(targetHeader, " - ", 2)[0]
	changelogSection, found := extractSectionFromChangelog(string(changelogContent), headerPrefix)
	if !found {
		r := validationResult{
			Valid:    false,
			Errors:   []string{fmt.Sprintf("Section %q not found in %s", targetHeader, changelogPath)},
			Warnings: []string{},
		}
		writeResult(r)
		os.Exit(1)
	}

	result := validateProvenance(validateProvenanceParams{
		evidence:         &evidence,
		provenance:       &provenance,
		changelogSection: changelogSection,
		expectedHeader:   targetHeader,
	})

	writeResult(result)

	if !result.Valid {
		fmt.Fprintf(os.Stderr, "Validation failed: %d error(s)\n", len(result.Errors))
		os.Exit(1)
	}

	if len(result.Warnings) > 0 {
		fmt.Fprintf(os.Stderr, "Validation passed (%d warning(s))\n", len(result.Warnings))
	} else {
		fmt.Fprintln(os.Stderr, "Validation passed")
	}
}
